packages/liveblocks-core/src/connection.ts

```diff
@@ -0,0 +1,756 @@
+
+  function log(...args: unknown[]) {
+    // eslint-disable-next-line
+    console.warn(
+      `${((new Date().getTime() - start) / 1000).toFixed(2)} [FSM #${fsm.id}]`,
+      ...args
+    );
+  }
+  const unsubs = [
+    fsm.events.didReceiveEvent.subscribe((e) => {
+      log(`Event ${e.type}`);
+    }),
+    fsm.events.willTransition.subscribe(({ from, to }) => {
+      log("Transitioning", from, "→", to);
+    }),
+    fsm.events.didPatchContext.subscribe((patch) => {
+      log(`Patched: ${JSON.stringify(patch)}`);
+      // log(`New context: ${JSON.stringify(fsm.context)}`);
+    }),
+    fsm.events.didIgnoreEvent.subscribe((e) => {
+      log("Ignored event", e, "(current state won't handle it)");
+    }),
+    // fsm.events.willExitState.subscribe((s) => {
+    //   log("Exiting state", s);
+    // }),
+    // fsm.events.didEnterState.subscribe((s) => {
+    //   log("Entering state", s);
+    // }),
+  ];
+  return () => {
+    for (const unsub of unsubs) {
+      unsub();
+    }
+  };
+}
+
+function defineConnectivityEvents(fsm: FSM<Context, Event, State>): {
+  statusDidChange: Observable<PublicConnectionStatus>;
+  didConnect: Observable<void>;
+  didDisconnect: Observable<void>;
+} {
+  // Emitted whenever a new WebSocket connection attempt suceeds
+  const statusDidChange = makeEventSource<PublicConnectionStatus>();
+  const didConnect = makeEventSource<void>();
+  const didDisconnect = makeEventSource<void>();
+
+  let oldPublicStatus: PublicConnectionStatus | null = null;
+
+  fsm.events.didEnterState.subscribe((newState) => {
+    const newPublicStatus = toPublicConnectionStatus(newState);
+    statusDidChange.notify(newPublicStatus);
+
+    if (oldPublicStatus === "open" && newPublicStatus !== "open") {
+      didDisconnect.notify();
+    } else if (oldPublicStatus !== "open" && newPublicStatus === "open") {
+      didConnect.notify();
+    }
+    oldPublicStatus = newPublicStatus;
+  });
+
+  return {
+    statusDidChange: statusDidChange.observable,
+    didConnect: didConnect.observable,
+    didDisconnect: didDisconnect.observable,
+  };
+}
+
+function createStateMachine<T extends BaseAuthResult>(delegates: Delegates<T>) {
+  // Create observable event sources, which this machine will call into when
+  // specific events happen
+  const onMessage = makeEventSource<IWebSocketMessageEvent>();
+
+  // Emitted whenever the server deliberately closes the connection for
+  // a specific Liveblocks reason
+  const onLiveblocksError = makeEventSource<LiveblocksError>();
+
+  const initialContext: Context & { token: T | null } = {
+    token: null,
+    socket: null,
+
+    // Bumped to the next "tier" every time a connection attempt fails (no matter
+    // whether this is for the authentication server or the websocket server).
+    // Reset every time a connection succeeded.
+    backoffDelay: LOW_DELAY,
+    // numRetries: 0,
+  };
+
+  const fsm = new FSM<Context, Event, State>(initialContext)
```

Comment:
could we find a more descriptive name for the state machine? maybe connectionState or something?

Reply:
Yeah, good point. I've renamed it from `fsm` → `machine` now in 5d1955272c4353ce72bd10f85b0c44255bd69404, and documented its purpose where it's first introduced.

I tried `connectionState` and some other longer/more verbose names, but it makes the code less readable IMHO, mostly because the machine instance isn't that important. The attention should go to all the transitions being configured there.

(A bit weird maybe, but it feels a bit like how you wouldn't name a `req`/`res` arg in a request handler `createRoomReq`/`createRoomRes` either—longer names would make it less readable because it draws attention from what's really important.)

Either way, I agree `fsm` wasn't a good variable name—hope this helps!